Compilation of `trilinos_tpetra_sparsity_pattern` fails when `DEAL_II_UNITY_BUILD=ON`

Tpetra related classes frequently use locally defined types (usually `size_type`, `MapType` and `GraphType`) within anonymous namespaces to define implementations of various methods, like here:

https://github.com/dealii/dealii/blob/be64a602e63f85202d8738a4642bea6369f8776e/source/lac/trilinos_tpetra_sparsity_pattern.cc#L261-L276

This approach works fine provided those files with such definitions are compiled as separate units. However if `DEAL_II_UNITY_BUILD=ON`, then this is no longer the case and conflicting type definitions can get into the same unit, this then obviously leads to the following compilcations errors:

```
In file included from ./dealii-fork-build/source/lac/unity_2.cc:8:
./dealii-fork/source/lac/trilinos_tpetra_sparsity_pattern.cc:264:73: error: ‘template<class MemorySpace> using size_type = typename dealii::LinearAlgebra::TpetraWrappers::SparsityPattern::size_type’ redeclared as different kind of entity
  264 |       using size_type = typename SparsityPattern<MemorySpace>::size_type;
      |                                                                         ^
In file included from ./dealii-fork/source/lac/trilinos_tpetra_sparse_matrix.cc:19,
                 from ./dealii-fork-build/source/lac/unity_2.cc:7:
./dealii-fork/include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h:145:13: note: previous declaration ‘using size_type = using signed_global_dof_index = long long int’
  145 |       using size_type = dealii::types::signed_global_dof_index;
      |             ^~~~~~~~~
In file included from ./dealii-fork-build/source/lac/unity_2.cc:8:
./dealii-fork/source/lac/trilinos_tpetra_sparsity_pattern.cc:267:13: error: conflicting declaration of template ‘template<class MemorySpace> using MapType = Tpetra::Map<int, long long int, typename dealii::LinearAlgebra::TpetraWrappers::SparsityPattern<MemorySpace>::NodeType>’
  267 |       using MapType =
      |             ^~~~~~~
In file included from ./dealii-fork/source/lac/trilinos_tpetra_sparse_matrix.cc:19,
                 from ./dealii-fork-build/source/lac/unity_2.cc:7:
./dealii-fork/include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h:148:13: note: previous declaration ‘template<class NodeType> using MapType = Tpetra::Map<int, long long int, NodeType>’
  148 |       using MapType =
      |             ^~~~~~~
In file included from ./dealii-fork-build/source/lac/unity_2.cc:8:
./dealii-fork/source/lac/trilinos_tpetra_sparsity_pattern.cc:273:13: error: conflicting declaration of template ‘template<class MemorySpace> using GraphType = Tpetra::CrsGraph<int, long long int, typename dealii::LinearAlgebra::TpetraWrappers::SparsityPattern<MemorySpace>::NodeType>’
  273 |       using GraphType =
      |             ^~~~~~~~~
In file included from ./dealii-fork/source/lac/trilinos_tpetra_sparse_matrix.cc:19,
                 from ./dealii-fork-build/source/lac/unity_2.cc:7:
./dealii-fork/include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h:159:13: note: previous declaration ‘template<class NodeType> using GraphType = Tpetra::CrsGraph<int, long long int, NodeType>’
  159 |       using GraphType =
      |             ^~~~~~~~~
```

Apparently, merging different `.cc` files may break the original intentions of a developer, so what are possible solutions coming to my mind:
1. Use different alias names within such anonymous namespaces (not very elegant and error prone).
2. Make in cmake a list of files exceptions which should not be merged into joint compilations units.

--- a/include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h
+++ b/include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h
@@ -140,7 +140,7 @@ namespace LinearAlgebra
 
 
     // reinit_matrix():
-    namespace
+    namespace SparseMatrixImpl
     {
       using size_type = dealii::types::signed_global_dof_index;
 
@@ -400,7 +400,7 @@ namespace LinearAlgebra
           Utilities::Trilinos::internal::make_rcp<MatrixType<Number, NodeType>>(
             graph);
       }
-    } // namespace
+    } // namespace SparseMatrixImpl
 
 
 
@@ -530,7 +530,7 @@ namespace LinearAlgebra
     SparseMatrix<Number, MemorySpace>::reinit(
       const SparsityPatternType &sparsity_pattern)
     {
-      reinit_matrix<Number, NodeType, SparsityPatternType>(
+      SparseMatrixImpl::reinit_matrix<Number, NodeType, SparsityPatternType>(
         complete_index_set(sparsity_pattern.n_rows()),
         complete_index_set(sparsity_pattern.n_cols()),
         sparsity_pattern,
@@ -675,7 +675,7 @@ namespace LinearAlgebra
       const MPI_Comm             communicator,
       const bool                 exchange_data)
     {
-      reinit_matrix<Number, NodeType, SparsityPatternType>(
+      SparseMatrixImpl::reinit_matrix<Number, NodeType, SparsityPatternType>(
         row_parallel_partitioning,
         col_parallel_partitioning,
         sparsity_pattern,

--- a/source/lac/trilinos_tpetra_sparsity_pattern.cc
+++ b/source/lac/trilinos_tpetra_sparsity_pattern.cc
@@ -258,7 +258,7 @@ namespace LinearAlgebra
 
 
 
-    namespace
+    namespace SparsityPatternImpl
     {
       template <typename MemorySpace>
       using size_type = typename SparsityPattern<MemorySpace>::size_type;
@@ -465,7 +465,7 @@ namespace LinearAlgebra
 
         graph->globalAssemble();
       }
-    } // namespace
+    } // namespace SparsityPatternImpl
 
 
     template <typename MemorySpace>
@@ -478,7 +478,7 @@ namespace LinearAlgebra
                                   parallel_partitioning.size());
       Teuchos::RCP<MapType> map =
         parallel_partitioning.make_tpetra_map_rcp(communicator, false);
-      reinit_sp<MemorySpace>(
+      SparsityPatternImpl::reinit_sp<MemorySpace>(
         map, map, n_entries_per_row, column_space_map, graph, nonlocal_graph);
     }
 
@@ -495,7 +495,7 @@ namespace LinearAlgebra
                                   parallel_partitioning.size());
       Teuchos::RCP<MapType> map =
         parallel_partitioning.make_tpetra_map_rcp(communicator, false);
-      reinit_sp<MemorySpace>(
+      SparsityPatternImpl::reinit_sp<MemorySpace>(
         map, map, n_entries_per_row, column_space_map, graph, nonlocal_graph);
     }
 
@@ -515,12 +515,12 @@ namespace LinearAlgebra
         row_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
       Teuchos::RCP<MapType> col_map =
         col_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
-      reinit_sp<MemorySpace>(row_map,
-                             col_map,
-                             n_entries_per_row,
-                             column_space_map,
-                             graph,
-                             nonlocal_graph);
+      SparsityPatternImpl::reinit_sp<MemorySpace>(row_map,
+                                                  col_map,
+                                                  n_entries_per_row,
+                                                  column_space_map,
+                                                  graph,
+                                                  nonlocal_graph);
     }
 
 
@@ -539,12 +539,12 @@ namespace LinearAlgebra
         row_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
       Teuchos::RCP<MapType> col_map =
         col_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
-      reinit_sp<MemorySpace>(row_map,
-                             col_map,
-                             n_entries_per_row,
-                             column_space_map,
-                             graph,
-                             nonlocal_graph);
+      SparsityPatternImpl::reinit_sp<MemorySpace>(row_map,
+                                                  col_map,
+                                                  n_entries_per_row,
+                                                  column_space_map,
+                                                  graph,
+                                                  nonlocal_graph);
     }
 
 
@@ -564,12 +564,12 @@ namespace LinearAlgebra
         row_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
       Teuchos::RCP<MapType> col_map =
         col_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
-      reinit_sp<MemorySpace>(row_map,
-                             col_map,
-                             n_entries_per_row,
-                             column_space_map,
-                             graph,
-                             nonlocal_graph);
+      SparsityPatternImpl::reinit_sp<MemorySpace>(row_map,
+                                                  col_map,
+                                                  n_entries_per_row,
+                                                  column_space_map,
+                                                  graph,
+                                                  nonlocal_graph);
 
       IndexSet nonlocal_partitioner = writable_rows;
       AssertDimension(nonlocal_partitioner.size(),
@@ -615,13 +615,14 @@ namespace LinearAlgebra
         row_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
       Teuchos::RCP<MapType> col_map =
         col_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
-      reinit_sp<SparsityPatternType, MemorySpace>(row_map,
-                                                  col_map,
-                                                  nontrilinos_sparsity_pattern,
-                                                  exchange_data,
-                                                  column_space_map,
-                                                  graph,
-                                                  nonlocal_graph);
+      SparsityPatternImpl::reinit_sp<SparsityPatternType, MemorySpace>(
+        row_map,
+        col_map,
+        nontrilinos_sparsity_pattern,
+        exchange_data,
+        column_space_map,
+        graph,
+        nonlocal_graph);
     }
 
 
@@ -643,13 +644,14 @@ namespace LinearAlgebra
                                   parallel_partitioning.size());
       Teuchos::RCP<MapType> map =
         parallel_partitioning.make_tpetra_map_rcp(communicator, false);
-      reinit_sp<SparsityPatternType, MemorySpace>(map,
-                                                  map,
-                                                  nontrilinos_sparsity_pattern,
-                                                  exchange_data,
-                                                  column_space_map,
-                                                  graph,
-                                                  nonlocal_graph);
+      SparsityPatternImpl::reinit_sp<SparsityPatternType, MemorySpace>(
+        map,
+        map,
+        nontrilinos_sparsity_pattern,
+        exchange_data,
+        column_space_map,
+        graph,
+        nonlocal_graph);
     }
 
 
@@ -701,7 +703,7 @@ namespace LinearAlgebra
           0,
           Utilities::Trilinos::tpetra_comm_self());
 
-      reinit_sp<SparsityPatternType, MemorySpace>(
+      SparsityPatternImpl::reinit_sp<SparsityPatternType, MemorySpace>(
         rows, columns, sp, false, column_space_map, graph, nonlocal_graph);
     }
 

